### PR TITLE
MVP Multi rdb load

### PR DESCRIPTION
**Pull request by dranikpg**

### CODE REVIEW

File: src/server/rdb_save.cc [hunk 1]
```diff
@@ -743,6 +743,9 @@ class RdbSaver::Impl {
 
   error_code ConsumeChannel();
 
+  // Manually close channel.
+  void CloseChannel();
+
```
Comment (dranikpg) [inline src/server/rdb_save.cc:748]: Forgot to remove this... If we have no producers (header case), the channel should be closed on start (I've checked the internals). See the DCHECK below on github

File: src/server/server_family.cc [hunk 2]
```diff
@@ -666,6 +722,7 @@ void ServerFamily::StatsMC(std::string_view section, facade::ConnectionContext*
 static void RunStage(bool new_version, std::function<void(unsigned)> cb) {
   if (new_version) {
     shard_set->RunBlockingInParallel([&](EngineShard* es) { cb(es->shard_id()); });
+    cb(shard_set->size());
```
Comment (dranikpg) [inline src/server/server_family.cc:725]: TODO: Add comment this last odd call if for the summary file

File: src/server/server_family.cc [hunk 3]
```diff
@@ -732,24 +789,45 @@ error_code ServerFamily::DoSave(bool new_version, Transaction* trans, string* er
   };
 
   if (new_version) {
-    snapshots.resize(shard_set->size());
+    snapshots.resize(shard_set->size() + 1);
+
+    // save header file
+    {
+      error_code local_ec;
```
Comment (dranikpg) [inline src/server/server_family.cc:796]: DoSave is a bit big now... Maybe split it up?

File: src/server/server_family.cc [hunk 4]
```diff
@@ -375,22 +386,47 @@ void ServerFamily::Shutdown() {
...
 
   LOG(INFO) << "Loading " << load_path;
 
   GlobalState new_state = service_.SwitchState(GlobalState::ACTIVE, GlobalState::LOADING);
   if (new_state != GlobalState::LOADING) {
     LOG(WARNING) << GlobalStateName(new_state) << " in progress, ignored";
-    return;
+    // TODO: What to return?? -> is_valid check on receiver?
+    return {};
   }
```
Comment (dranikpg) [inline src/server/server_family.cc:430]: The future is invalid in this case, but I see no other way

File: src/server/dflycmd.cc [hunk 5]
```diff
@@ -317,7 +318,7 @@ OpStatus DflyCmd::FullSyncInShard(uint32_t syncid, Transaction* t, EngineShard*
   Connection* conn = shard_it->second.conn;
   lk.unlock();
 
-  unique_ptr<RdbSaver> saver = make_unique<RdbSaver>(conn->socket(), true /* single shard */,
+  unique_ptr<RdbSaver> saver = make_unique<RdbSaver>(conn->socket(), SaveMode::RDB /* single shard */,
                                                      false /* do not align writes */);
```
[the remark this answers is not in the document]
Comment (romange) [reply src/server/dflycmd.cc:322]: currently, replication has a control channel and N data channels. We can use the control channel to transmit the "summary". We can also use one of the shards as well. I find the "summary" approach cleaner than using , say shard 0 for transmitting the meta-data but maybe shard0 is simpler?

File: src/server/rdb_save.h [hunk 6]
```diff
@@ -67,14 +67,21 @@ class AlignedBuffer : public ::io::Sink {
   off_t buf_offs_ = 0;
 };
 
+// SaveMode for snapshot.
+enum class SaveMode {
+  SUMMARY,       // Save only header values (summary.dfs).
```
Comment (romange) [inline src/server/rdb_save.h:72]: it seems that you already decided.

File: src/server/rdb_save.cc [hunk 7]
```diff
@@ -864,10 +867,15 @@ void RdbSaver::Impl::FillFreqMap(RdbTypeFreqMap* dest) const {
   }
 }
 
-RdbSaver::RdbSaver(::io::Sink* sink, bool single_shard, bool align_writes) {
+RdbSaver::RdbSaver(::io::Sink* sink, SaveMode save_mode, bool align_writes) {
   CHECK_NOTNULL(sink);
 
-  impl_.reset(new Impl(align_writes, single_shard ? 1 : shard_set->size(), sink));
+  int size = 0;
+  if (save_mode == SaveMode::SINGLE_SHARD) size = 1;
```
Comment (romange) [inline src/server/rdb_save.cc:874]: I do not like `if else if` one-liners.
using switch construct will at least cause compiler error if one of the enums is not covered.

**Review by romange — COMMENTED** [review]
few comments. I do not feel strongly about any of them, just wanted to hear your thoughts.

File: src/server/server_family.cc [hunk 8]
```diff
@@ -405,15 +440,48 @@ void ServerFamily::Load(const std::string& load_path) {
...
+      if (ec && !ec_result)
+        ec_result = ec;
+    }
+    VLOG(1) << "Load finished";
+    service_.SwitchState(GlobalState::LOADING, GlobalState::ACTIVE);
+    ec_promise.set_value(ec_result);
+  };
+  pool.GetNextProactor()->LaunchFiber(std::move(load_join_fiber)).detach();
+
+  return ec_future;
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/server_family.cc:484]: I gives me the possibility to return a delayed value from Load - the error code. This is needed in dfycmd, main_service just joins before shutdown. What other way could this be done?

File: src/server/server_family.cc [hunk 9]
```diff
@@ -405,15 +440,48 @@ void ServerFamily::Load(const std::string& load_path) {
...
-      shard_count() < pool.size() ? pool.at(shard_count()) : pool.GetNextProactor();
-
-  load_fiber_ = proactor->LaunchFiber([load_path, this] {
-    auto ec = LoadRdb(load_path);
-    LOG_IF(ERROR, ec) << "Error loading file " << ec.message();
-  });
+
+  // Channel to receive error codes from loading fibers.
+  // buffer size is no less than there are load fibers to prevent any blocking.
+  auto chan = std::make_shared<::util::fibers_ext::SimpleChannel<std::error_code>>(paths.size());
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/server_family.cc:446]: > if you would explicitly keep fibers handles in an array (instead of detaching) you could join on them inside load_join_fiber and get the same outcome

I did this first. But then I realized I need each fibers return value (the error code).

> instead of just using fibers::mutex
I've seen this in your code. I even fixed a bug with the same pattern (but opstatus instead of ec) 😄 

I not really fond of it. First, its error prone (see my comment above). Second, its kind of very simple and limited. No info what else happened, where it originated. In case you wanted to add more debug messages or report the corrupted file, you'd have to do that further down in parsing code. In case you want to detach the fiber after the first error - you can't either.

But I have no problem with having your suggested approach here.

> an opportunity to introduce AggregateError thread-safe class that merges into itself std::error_code if it's set
That seems like a helpful utility. At least you don't have to setup the mutex and overwrite-with-0 checks yourself. 

By the way, the same applies to OpStatus. What is more, its just an enum - it can be edited atomically without any mutex at all

File: src/server/debugcmd.cc [hunk 10]
```diff
@@ -211,9 +211,8 @@ void DebugCmd::Load(string_view filename) {
     path = dir_path;
   }
 
-  // switches back to
-  ec = sf_.LoadRdb(path.generic_string());
-  if (ec) {
+  auto fut_ec = sf_.Load(path.generic_string());
+  if (fut_ec.valid() && !(ec = fut_ec.get())) {
```
Comment (romange) [inline src/server/debugcmd.cc:215]: I would rather avoid using `=` inside conditions.

File: src/server/server_family.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (romange) [reply src/server/server_family.cc:446]: Sounds good

File: src/server/utility.h [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+#include <boost/fiber/mutex.hpp>
+
+#include "server/common.h"
+
+namespace dfly {
+
+// AggregateValue is a thread safe utility to store the first
+// non-default value.
+template <typename T> struct AggregateValue {
```
Comment (dranikpg) [inline src/server/utility.h:15]: Maybe server/utility.h is not the best place for this helper. My other options were common and error, but they seem to specialized for utils

**Comment by romange** [comment]
My usual place to stuff such things is common.h . I am all right with
utility if you tell me what goes there and what to common :)

**Comment by dranikpg** [comment]
Half of common are usings/forward decls/types. The other half are utils... If its okay I'll place it there, its just that common.h is also included in most headers.
A to-do for future would be splitting common into utils (real utils) and common (types).

File: src/server/server_family.cc [hunk 12]
```diff
@@ -732,24 +803,46 @@ error_code ServerFamily::DoSave(bool new_version, Transaction* trans, string* er
...
-    // In the new version we open a file per shard
+    // Save summary file.
+    {
+      error_code local_ec;
+
+      fs::path summary_file = filename, abs_path = path;
+      ExtendFilename(now, "summary", &summary_file);
+      abs_path += summary_file;
+
+      auto res = uring::OpenLinux(abs_path.generic_string(), kRdbWriteFlags, 0666);
```
[the remark this answers is not in the document]
Comment (dranikpg) [reply src/server/server_family.cc:818]: I'm merging into the ReplicationFullSync branch, there is no epoll yet

File: src/server/server_family.cc [hunk 13]
```diff
@@ -375,22 +387,45 @@ void ServerFamily::Shutdown() {
   });
 }
 
-void ServerFamily::Load(const std::string& load_path) {
-  CHECK(!load_fiber_.get_id());
+// Load starts as many fibers as there are files and collects
+// the load results in a channel.
```
Comment (romange) [inline src/server/server_family.cc:391]: the comment is wrong now.

File: src/server/server_family.cc [hunk 14]
```diff
@@ -41,6 +43,7 @@ extern "C" {
 #include "server/version.h"
 #include "strings/human_readable.h"
 #include "util/accept_server.h"
+#include "util/fibers/simple_channel.h"
```
Comment (romange) [inline src/server/server_family.cc:46]: probably not needed anymore

File: src/server/server_family.cc [hunk 15]
```diff
@@ -405,15 +440,47 @@ void ServerFamily::Load(const std::string& load_path) {
...
+  }
+
+  boost::fibers::promise<std::error_code> ec_promise;
+  boost::fibers::future<std::error_code> ec_future = ec_promise.get_future();
+
+  // Run fiber that empties the channel and sets ec_promise.
+  auto load_join_fiber = [this, first_error, load_fibers = std::move(load_fibers),
+                          ec_promise = std::move(ec_promise)]() mutable {
+    for (auto& fiber : load_fibers) {
+      if (fiber.joinable()) {
```
Comment (romange) [inline src/server/server_family.cc:472]: they are all joinable, imho. so it should always be true

File: src/server/server_family.cc [hunk 16]
```diff
@@ -405,15 +440,47 @@ void ServerFamily::Load(const std::string& load_path) {
...
+        fiber.join();
+      }
+    }
+
+    VLOG(1) << "Load finished";
+    service_.SwitchState(GlobalState::LOADING, GlobalState::ACTIVE);
+    ec_promise.set_value(**first_error);
+  };
+  pool.GetNextProactor()->LaunchFiber(std::move(load_join_fiber)).detach();
```
Comment (romange) [inline src/server/server_family.cc:481]: `Dispatch(...)` is equivalent to `LaunchFiber(...).detach();`

**Review by romange — APPROVED** [review]

**Review by romange — APPROVED** [review]
